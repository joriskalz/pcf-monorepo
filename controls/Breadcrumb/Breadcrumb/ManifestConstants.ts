export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Clickable = 'ItemClickable',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
