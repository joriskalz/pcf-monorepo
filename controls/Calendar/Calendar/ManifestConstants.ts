export const enum InputEvents {
    SetFocus = 'SetFocus',
}
