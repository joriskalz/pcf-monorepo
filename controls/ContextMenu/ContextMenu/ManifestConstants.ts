export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    ParentKey = 'ItemParentKey',
    IconOnly = 'ItemIconOnly',
    Visible = 'ItemVisible',
    Checked = 'ItemChecked',
    ItemHeader = 'ItemHeader',
    ItemTopDivider = 'ItemTopDivider',
    ItemDivider = 'ItemDivider',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
}

export const TextAlignmentTypes = {
    '0': 'center',
    '1': 'left',
    '2': 'right',
};
