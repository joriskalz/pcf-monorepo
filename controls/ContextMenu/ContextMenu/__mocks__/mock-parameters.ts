/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockTwoOptionsProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        InputEvent: new MockStringProperty(),
        Theme: new MockStringProperty(),
        TextAlignment: new MockEnumProperty(),
        items: new MockDataSet([]),
        Chevron: new MockTwoOptionsProperty(),
        FillColor: new MockStringProperty(),
        BorderColor: new MockStringProperty(),
        BorderRadius: new MockWholeNumberProperty(),
        FontColor: new MockStringProperty(),
        FontSize: new MockWholeNumberProperty(),
        HoverBorderColor: new MockStringProperty(),
        HoverFillColor: new MockStringProperty(),
        HoverFontColor: new MockStringProperty(),
        HoverIconColor: new MockStringProperty(),
        IconColor: new MockStringProperty(),
        IconSize: new MockWholeNumberProperty(),
    };
}
