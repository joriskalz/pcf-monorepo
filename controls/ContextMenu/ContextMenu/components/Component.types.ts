export interface CanvasContextMenuItem {
    id: string;
    enabled?: boolean;
    name: string;
    iconName?: string;
    iconColor?: string;
    itemCount?: number;
    iconOnly?: boolean;
    visible?: boolean;
    parentItemKey?: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
    checked: boolean;
    isHeader?: boolean;
    divider?: boolean;
    topDivider?: boolean;
}

export interface CanvasContextMenuProps {
    width?: number;
    height?: number;
    text?: string;
    textColor?: string;
    iconSize?: number;
    iconColor?: string;
    hoverIconColor?: string;
    fontSize?: number;
    fontColor?: string;
    hoverFontColor?: string;
    borderColor?: string;
    hoverBorderColor?: string;
    borderRadius?: number;
    fillColor?: string;
    hoverFillColor?: string;
    showChevron?: boolean;
    justify?: 'center' | 'left' | 'right';
    items: CanvasContextMenuItem[];
    onSelected: (item?: CanvasContextMenuItem) => void;
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    setFocus?: string;
}
