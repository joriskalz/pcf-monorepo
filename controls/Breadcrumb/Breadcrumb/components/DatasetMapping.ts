import { IBreadcrumbItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomBreadcrumbItem } from './components.types';

export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomBreadcrumbItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            text: record.getFormattedValue(ItemColumns.DisplayName),
            clickable: record.getValue(ItemColumns.Clickable) as boolean,
            data: record,
        } as ICustomBreadcrumbItem;
    });
}

export function getBreadcrumbItems(
    items: ICustomBreadcrumbItem[],
    onClick: (ev?: React.MouseEvent<HTMLElement>, item?: ICustomBreadcrumbItem) => void,
): IBreadcrumbItem[] {
    return items.map((item) =>
        item.clickable === false
            ? {
                  id: item.id,
                  key: item.key,
                  text: item.text,
              }
            : {
                  id: item.id,
                  key: item.key,
                  text: item.text,
                  onClick: onClick,
              },
    ) as IBreadcrumbItem[];
}

function getDummyAction(key: string): ICustomBreadcrumbItem {
    return {
        id: key,
        key: key,
        text: 'Item ' + key,
        clickable: true,
    } as ICustomBreadcrumbItem;
}
