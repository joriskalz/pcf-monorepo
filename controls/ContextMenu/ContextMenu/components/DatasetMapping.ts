import { CanvasContextMenuItem } from './Component.types';
import {
    ContextualMenuItemType,
    getColorFromString,
    getShade,
    IButtonStyles,
    IContextualMenuItem,
    IIconProps,
    Shade,
} from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
const CSS_IMPORTANT = ' !important';
export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasContextMenuItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = (record.getFormattedValue(ItemColumns.Key) as string) ?? '';
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getFormattedValue(ItemColumns.DisplayName),
            enabled: (record.getValue(ItemColumns.Enabled) as boolean) ?? undefined,
            iconName: record.getFormattedValue(ItemColumns.IconName),
            iconColor: record.getFormattedValue(ItemColumns.IconColor),
            parentItemKey: record.getFormattedValue(ItemColumns.ParentKey),
            iconOnly: (record.getValue(ItemColumns.IconOnly) as boolean) ?? undefined,
            checked: (record.getValue(ItemColumns.Checked) as boolean) ?? undefined,
            visible: (record.getValue(ItemColumns.Visible) as boolean) ?? undefined,
            isHeader: (record.getValue(ItemColumns.ItemHeader) as boolean) ?? undefined,
            topDivider: (record.getValue(ItemColumns.ItemTopDivider) as boolean) ?? undefined,
            divider: (record.getValue(ItemColumns.ItemDivider) as boolean) ?? undefined,
            data: record,
        } as CanvasContextMenuItem;
    });
}

function getDummyAction(key: string): CanvasContextMenuItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasContextMenuItem;
}

export function getCommandsWithChildren(
    items: CanvasContextMenuItem[],
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasContextMenuItem[],
    item: CanvasContextMenuItem,
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i !== item && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
    const buttonStyles = {
        root: {
            background: 'rgba(255, 255, 255,0)',
        },
    } as IButtonStyles;

    const iconProps =
        item.iconName &&
        ({
            iconName: item.iconName,
        } as IIconProps);

    // If custom icon color, find the shade slots for the hover/pressed state
    if (iconProps && item.iconColor) {
        const iconColor = item.iconColor;
        iconProps.styles = { root: { color: item.iconColor + CSS_IMPORTANT } };

        // Attempt to set a darker shade for hover/pressed colors
        const primaryColor = getColorFromString(iconColor);
        if (primaryColor) {
            const iconColorDarker = getShade(primaryColor, Shade.Shade8);
            const secondaryColorStyle = { color: iconColorDarker?.str + CSS_IMPORTANT };
            buttonStyles.iconHovered = secondaryColorStyle;
            buttonStyles.iconPressed = secondaryColorStyle;
            buttonStyles.iconExpanded = secondaryColorStyle;
            buttonStyles.iconChecked = secondaryColorStyle;
            buttonStyles.iconExpandedHovered = secondaryColorStyle;
        }
    }

    const props = {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: iconProps,
        canCheck: item.checked !== undefined,
        checked: item.checked,
        data: item,
        buttonStyles: buttonStyles,
    } as IContextualMenuItem;

    if (item.isHeader === true && subItems.length > 0) {
        // Semantic Section
        props.itemType = ContextualMenuItemType.Section;
        props.sectionProps = {
            title: item.name,
            topDivider: item.topDivider,
            bottomDivider: item.divider,
            items: subItems,
        };
    } else if (item.isHeader === true) {
        // Simple Section Break
        props.itemType = ContextualMenuItemType.Header;
    } else if (item.divider === true) {
        // Divider Break
        props.itemType = ContextualMenuItemType.Divider;
    } else {
        props.itemType = ContextualMenuItemType.Normal;
        props.subMenuProps =
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined;
    }
    return props;
}
