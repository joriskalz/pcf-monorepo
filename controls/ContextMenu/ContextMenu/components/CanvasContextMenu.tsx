import * as React from 'react';
import { CanvasContextMenuItem, CanvasContextMenuProps } from './Component.types';
import { useAsync } from '@fluentui/react-hooks';
import {
    createTheme,
    IPartialTheme,
    ThemeProvider,
    DefaultButton,
    IButton,
    IButtonProps,
    IContextualMenuItem,
    IButtonStyles,
} from '@fluentui/react';
import { getCommandsWithChildren } from './DatasetMapping';

export const CanvasContextMenu = React.memo((props: CanvasContextMenuProps) => {
    const { items, onSelected, disabled, setFocus, themeJSON, showChevron, tabIndex } = props;

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    const componentRef = React.useRef<IButton>(null);
    const async = useAsync();
    React.useEffect(() => {
        if (setFocus && setFocus !== '' && componentRef) {
            async.requestAnimationFrame(() => {
                (componentRef as React.RefObject<IButton>).current?.focus();
            });
        }
    }, [setFocus, componentRef, async]);

    const onClick = React.useCallback(
        (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => {
            // If this is a check item, then prevent closing the menu
            if (item?.canCheck) ev?.preventDefault();
            onSelected(item?.data);
            return true;
        },
        [onSelected],
    );

    const buttonProps = React.useMemo(() => {
        const props = {} as Partial<IButtonProps>;
        if (items && items.length > 0) {
            const firstItem = items[0];
            props.iconProps = {
                iconName: firstItem.iconName,
            };
            props.disabled = disabled === true;
            props.text = firstItem.iconOnly ? '' : firstItem.name;
            props.onSelect = () => {
                // Since you can't raise the OnSelect event with no Selected Item
                // we use the first item to represent the root button click
                // so that it acts like a normal Button
                onSelected(firstItem);
            };
            const contextItems = getCommandsWithChildren(items.slice(1), disabled === true, onClick);
            if (contextItems && contextItems.length > 0) {
                props.menuProps = {
                    shouldFocusOnContainer: true,
                    items: contextItems,
                };
            }

            props.onRenderMenuIcon =
                showChevron === false
                    ? () => {
                          return null;
                      }
                    : undefined;

            return props;
        }
        return undefined;
    }, [disabled, items, onClick, onSelected, showChevron]);

    return (
        <ThemeProvider applyTo="none" theme={theme} className={'PowerCATFluentContextMenu'}>
            <DefaultButton
                {...buttonProps}
                styles={getButtonStyles(props, items)}
                componentRef={componentRef}
                tabIndex={tabIndex}
            />
        </ThemeProvider>
    );
});
CanvasContextMenu.displayName = 'CanvasContextMenu';

function getButtonStyles(props: CanvasContextMenuProps, items: CanvasContextMenuItem[]) {
    return React.useMemo(() => {
        const rootItem = items && items.length > 0 ? items[0] : undefined;
        const iconColor = props.iconColor ?? rootItem?.iconColor ?? props.fontColor;
        const textColor = props.fontColor;
        const hoverTextColor = props.hoverFontColor ?? textColor;
        const fillColor = props.fillColor;
        const hoverFillColor = props.hoverFillColor ?? fillColor;
        const hoverIconColor = props.hoverIconColor ?? props.hoverFontColor ?? props.iconColor ?? props.fontColor;
        return {
            root: {
                width: props.width,
                height: props.height,
                backgroundColor: fillColor,
                borderColor: props.borderColor,
                color: textColor,
                borderRadius: props.borderRadius,
                borderWidth: props.fillColor ? 1 : undefined,
                borderStyle: props.fillColor ? 'solid' : undefined,
                fontSize: props.fontSize,
            },
            rootHovered: {
                backgroundColor: hoverFillColor,
                borderColor: props.hoverBorderColor,
                color: hoverTextColor,
            },
            rootPressed: {
                backgroundColor: hoverFillColor,
                color: hoverTextColor,
            },
            rootExpanded: {
                backgroundColor: fillColor,
                color: textColor,
            },
            rootExpandedHovered: {
                backgroundColor: hoverFillColor,
                borderColor: props.hoverBorderColor,
                color: hoverTextColor,
            },
            icon: {
                color: iconColor,
                fontSize: props.iconSize ?? props.fontSize,
            },
            iconHovered: {
                color: hoverIconColor,
            },
            iconExpandedHovered: {
                color: hoverIconColor,
            },
            label: {
                textAlign: props.justify,
            },
        } as IButtonStyles;
    }, [
        items,
        props.borderColor,
        props.borderRadius,
        props.fillColor,
        props.fontColor,
        props.fontSize,
        props.height,
        props.hoverBorderColor,
        props.hoverFillColor,
        props.hoverFontColor,
        props.hoverIconColor,
        props.iconColor,
        props.iconSize,
        props.justify,
        props.width,
    ]);
}
