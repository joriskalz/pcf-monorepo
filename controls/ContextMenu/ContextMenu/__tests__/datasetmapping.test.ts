import { getCommandsWithChildren, getMenuItemsFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'commandOpen',
                [ItemColumns.DisplayName]: 'Open',
                [ItemColumns.IconName]: 'Open',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: true,
                [ItemColumns.IconOnly]: false,
            }),
            new MockEntityRecord('2', {
                [ItemColumns.Key]: 'commandNew',
                [ItemColumns.DisplayName]: 'New',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
            new MockEntityRecord('3', {
                [ItemColumns.Key]: 'commandSave',
                [ItemColumns.DisplayName]: 'Save',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
            }),
            // Sub Items First Level
            new MockEntityRecord('5', {
                [ItemColumns.Key]: 'commandNewFrom',
                [ItemColumns.DisplayName]: 'NewFrom',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandNew',
            }),
            new MockEntityRecord('6', {
                [ItemColumns.Key]: 'commandSaveAndClose',
                [ItemColumns.DisplayName]: 'SaveAndClose',
                [ItemColumns.IconName]: 'Save',
                [ItemColumns.IconColor]: 'green',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandSave',
            }),
            // Sub Items Second Level
            new MockEntityRecord('7', {
                [ItemColumns.Key]: 'commandNewFromSub1',
                [ItemColumns.DisplayName]: 'NewFromSub1',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandNewFrom',
            }),
            new MockEntityRecord('8', {
                [ItemColumns.Key]: 'commandNewFromSub2',
                [ItemColumns.DisplayName]: 'NewFromSub2',
                [ItemColumns.IconName]: 'New',
                [ItemColumns.IconColor]: 'blue',
                [ItemColumns.Enabled]: false,
                [ItemColumns.IconOnly]: true,
                [ItemColumns.ParentKey]: 'commandNewFrom',
            }),
        ];

        const onClickEvent = jest.fn();
        const actions = getMenuItemsFromDataset(new MockDataSet(items));
        const props = getCommandsWithChildren(actions, false, onClickEvent);
        expect(props).toMatchSnapshot();
    });
});
