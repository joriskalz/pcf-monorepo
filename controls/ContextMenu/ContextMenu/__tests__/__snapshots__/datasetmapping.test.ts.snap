// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "buttonStyles": Object {
      "root": Object {
        "background": "rgba(255, 255, 255,0)",
      },
    },
    "canCheck": false,
    "checked": undefined,
    "data": Object {
      "checked": undefined,
      "data": MockEntityRecord {
        "getNamedReference": [MockFunction],
        "id": "1",
        "values": Object {
          "ItemDisplayName": "Open",
          "ItemEnabled": true,
          "ItemIconColor": "blue",
          "ItemIconName": "Open",
          "ItemIconOnly": false,
          "ItemKey": "commandOpen",
        },
      },
      "divider": undefined,
      "enabled": true,
      "iconColor": "blue",
      "iconName": "Open",
      "iconOnly": false,
      "id": "1",
      "isHeader": undefined,
      "key": "commandOpen",
      "name": "Open",
      "parentItemKey": undefined,
      "topDivider": undefined,
      "visible": undefined,
    },
    "disabled": false,
    "iconOnly": false,
    "iconProps": Object {
      "iconName": "Open",
      "styles": Object {
        "root": Object {
          "color": "blue !important",
        },
      },
    },
    "itemType": 0,
    "key": "commandOpen",
    "onClick": [MockFunction],
    "subMenuProps": undefined,
    "text": "Open",
  },
  Object {
    "buttonStyles": Object {
      "root": Object {
        "background": "rgba(255, 255, 255,0)",
      },
    },
    "canCheck": false,
    "checked": undefined,
    "data": Object {
      "checked": undefined,
      "data": MockEntityRecord {
        "getNamedReference": [MockFunction],
        "id": "2",
        "values": Object {
          "ItemDisplayName": "New",
          "ItemEnabled": false,
          "ItemIconColor": "blue",
          "ItemIconName": "New",
          "ItemIconOnly": true,
          "ItemKey": "commandNew",
        },
      },
      "divider": undefined,
      "enabled": false,
      "iconColor": "blue",
      "iconName": "New",
      "iconOnly": true,
      "id": "2",
      "isHeader": undefined,
      "key": "commandNew",
      "name": "New",
      "parentItemKey": undefined,
      "topDivider": undefined,
      "visible": undefined,
    },
    "disabled": true,
    "iconOnly": true,
    "iconProps": Object {
      "iconName": "New",
      "styles": Object {
        "root": Object {
          "color": "blue !important",
        },
      },
    },
    "itemType": 0,
    "key": "commandNew",
    "onClick": [MockFunction],
    "subMenuProps": Object {
      "items": Array [
        Object {
          "buttonStyles": Object {
            "root": Object {
              "background": "rgba(255, 255, 255,0)",
            },
          },
          "canCheck": false,
          "checked": undefined,
          "data": Object {
            "checked": undefined,
            "data": MockEntityRecord {
              "getNamedReference": [MockFunction],
              "id": "5",
              "values": Object {
                "ItemDisplayName": "NewFrom",
                "ItemEnabled": false,
                "ItemIconColor": "blue",
                "ItemIconName": "New",
                "ItemIconOnly": true,
                "ItemKey": "commandNewFrom",
                "ItemParentKey": "commandNew",
              },
            },
            "divider": undefined,
            "enabled": false,
            "iconColor": "blue",
            "iconName": "New",
            "iconOnly": true,
            "id": "5",
            "isHeader": undefined,
            "key": "commandNewFrom",
            "name": "NewFrom",
            "parentItemKey": "commandNew",
            "topDivider": undefined,
            "visible": undefined,
          },
          "disabled": true,
          "iconOnly": true,
          "iconProps": Object {
            "iconName": "New",
            "styles": Object {
              "root": Object {
                "color": "blue !important",
              },
            },
          },
          "itemType": 0,
          "key": "commandNewFrom",
          "onClick": [MockFunction],
          "subMenuProps": Object {
            "items": Array [
              Object {
                "buttonStyles": Object {
                  "root": Object {
                    "background": "rgba(255, 255, 255,0)",
                  },
                },
                "canCheck": false,
                "checked": undefined,
                "data": Object {
                  "checked": undefined,
                  "data": MockEntityRecord {
                    "getNamedReference": [MockFunction],
                    "id": "7",
                    "values": Object {
                      "ItemDisplayName": "NewFromSub1",
                      "ItemEnabled": false,
                      "ItemIconColor": "blue",
                      "ItemIconName": "New",
                      "ItemIconOnly": true,
                      "ItemKey": "commandNewFromSub1",
                      "ItemParentKey": "commandNewFrom",
                    },
                  },
                  "divider": undefined,
                  "enabled": false,
                  "iconColor": "blue",
                  "iconName": "New",
                  "iconOnly": true,
                  "id": "7",
                  "isHeader": undefined,
                  "key": "commandNewFromSub1",
                  "name": "NewFromSub1",
                  "parentItemKey": "commandNewFrom",
                  "topDivider": undefined,
                  "visible": undefined,
                },
                "disabled": true,
                "iconOnly": true,
                "iconProps": Object {
                  "iconName": "New",
                  "styles": Object {
                    "root": Object {
                      "color": "blue !important",
                    },
                  },
                },
                "itemType": 0,
                "key": "commandNewFromSub1",
                "onClick": [MockFunction],
                "subMenuProps": undefined,
                "text": "NewFromSub1",
              },
              Object {
                "buttonStyles": Object {
                  "root": Object {
                    "background": "rgba(255, 255, 255,0)",
                  },
                },
                "canCheck": false,
                "checked": undefined,
                "data": Object {
                  "checked": undefined,
                  "data": MockEntityRecord {
                    "getNamedReference": [MockFunction],
                    "id": "8",
                    "values": Object {
                      "ItemDisplayName": "NewFromSub2",
                      "ItemEnabled": false,
                      "ItemIconColor": "blue",
                      "ItemIconName": "New",
                      "ItemIconOnly": true,
                      "ItemKey": "commandNewFromSub2",
                      "ItemParentKey": "commandNewFrom",
                    },
                  },
                  "divider": undefined,
                  "enabled": false,
                  "iconColor": "blue",
                  "iconName": "New",
                  "iconOnly": true,
                  "id": "8",
                  "isHeader": undefined,
                  "key": "commandNewFromSub2",
                  "name": "NewFromSub2",
                  "parentItemKey": "commandNewFrom",
                  "topDivider": undefined,
                  "visible": undefined,
                },
                "disabled": true,
                "iconOnly": true,
                "iconProps": Object {
                  "iconName": "New",
                  "styles": Object {
                    "root": Object {
                      "color": "blue !important",
                    },
                  },
                },
                "itemType": 0,
                "key": "commandNewFromSub2",
                "onClick": [MockFunction],
                "subMenuProps": undefined,
                "text": "NewFromSub2",
              },
            ],
          },
          "text": "NewFrom",
        },
      ],
    },
    "text": "New",
  },
  Object {
    "buttonStyles": Object {
      "root": Object {
        "background": "rgba(255, 255, 255,0)",
      },
    },
    "canCheck": false,
    "checked": undefined,
    "data": Object {
      "checked": undefined,
      "data": MockEntityRecord {
        "getNamedReference": [MockFunction],
        "id": "3",
        "values": Object {
          "ItemDisplayName": "Save",
          "ItemEnabled": false,
          "ItemIconColor": "green",
          "ItemIconName": "Save",
          "ItemIconOnly": true,
          "ItemKey": "commandSave",
        },
      },
      "divider": undefined,
      "enabled": false,
      "iconColor": "green",
      "iconName": "Save",
      "iconOnly": true,
      "id": "3",
      "isHeader": undefined,
      "key": "commandSave",
      "name": "Save",
      "parentItemKey": undefined,
      "topDivider": undefined,
      "visible": undefined,
    },
    "disabled": true,
    "iconOnly": true,
    "iconProps": Object {
      "iconName": "Save",
      "styles": Object {
        "root": Object {
          "color": "green !important",
        },
      },
    },
    "itemType": 0,
    "key": "commandSave",
    "onClick": [MockFunction],
    "subMenuProps": Object {
      "items": Array [
        Object {
          "buttonStyles": Object {
            "root": Object {
              "background": "rgba(255, 255, 255,0)",
            },
          },
          "canCheck": false,
          "checked": undefined,
          "data": Object {
            "checked": undefined,
            "data": MockEntityRecord {
              "getNamedReference": [MockFunction],
              "id": "6",
              "values": Object {
                "ItemDisplayName": "SaveAndClose",
                "ItemEnabled": false,
                "ItemIconColor": "green",
                "ItemIconName": "Save",
                "ItemIconOnly": true,
                "ItemKey": "commandSaveAndClose",
                "ItemParentKey": "commandSave",
              },
            },
            "divider": undefined,
            "enabled": false,
            "iconColor": "green",
            "iconName": "Save",
            "iconOnly": true,
            "id": "6",
            "isHeader": undefined,
            "key": "commandSaveAndClose",
            "name": "SaveAndClose",
            "parentItemKey": "commandSave",
            "topDivider": undefined,
            "visible": undefined,
          },
          "disabled": true,
          "iconOnly": true,
          "iconProps": Object {
            "iconName": "Save",
            "styles": Object {
              "root": Object {
                "color": "green !important",
              },
            },
          },
          "itemType": 0,
          "key": "commandSaveAndClose",
          "onClick": [MockFunction],
          "subMenuProps": undefined,
          "text": "SaveAndClose",
        },
      ],
    },
    "text": "Save",
  },
]
`;
